beater/interceptors/metadata.go

```diff
@@ -0,0 +1,83 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package interceptors
+
+import (
+	"context"
+	"net"
+	"net/http"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/peer"
+
+	"github.com/elastic/apm-server/utility"
+)
+
+// ClientMetadata returns an interceptor that intercepts unary gRPC requests,
+// extracts metadata relating to the gRPC client, and adds it to the context.
+//
+// Metadata can be extracted from context using ClientMetadataFromContext.
+func ClientMetadata() grpc.UnaryServerInterceptor {
+	return func(
+		ctx context.Context,
+		req interface{},
+		info *grpc.UnaryServerInfo,
+		handler grpc.UnaryHandler,
+	) (interface{}, error) {
+		var values ClientMetadataValues
+		if p, ok := peer.FromContext(ctx); ok {
+			if addr, ok := p.Addr.(*net.TCPAddr); ok {
+				values.SourceIP = addr.IP
+			}
+		}
+		if md, ok := metadata.FromIncomingContext(ctx); ok {
+			if ua := md["user-agent"]; len(ua) > 0 {
+				values.UserAgent = ua[0]
+			}
+			// Account for `forwarded`, `x-real-ip`, `x-forwarded-for` headers
+			if ip := utility.ExtractIPFromHeader(http.Header(md)); ip != nil {
+				values.SourceIP = ip
+			}
+		}
+		ctx = context.WithValue(ctx, clientMetadataKey{}, values)
+		return handler(ctx, req)
+	}
+}
+
+type clientMetadataKey struct{}
+
+// ContextWithClientMetadata returns a copy of ctx with values.
+func ContextWithClientMetadata(ctx context.Context, values ClientMetadataValues) context.Context {
+	return context.WithValue(ctx, clientMetadataKey{}, values)
```

Comment:
for my own education: using `clientMetadataKey{}` provides us with an identifier to extract the values, but it also doesn't allocate because it's an empty struct?

Reply:
Using an unexported struct type as the key is recommended by the Go docs (https://golang.org/pkg/context/):

>    // A key identifies a specific value in a Context. Functions that wish
>    // to store values in Context typically allocate a key in a global
>    // variable then use that key as the argument to context.WithValue and
>    // Context.Value. **A key can be any type that supports equality;
>    // packages should define keys as an unexported type to avoid
>    // collisions.**